src/xma/src/xmaapi/xmahw_hal.cpp

```diff
@@ -123,159 +124,72 @@ bool hal_configure(XmaHwCfg *hwcfg, XmaXclbinParameter *devXclbins, int32_t num_
-                } else {
-                    xma_logmsg(XMA_DEBUG_LOG, XMAAPI_MOD,"\tCU# %d - %s - default DDR bank:%d", d, (char*)tmp1.name, tmp1.default_ddr_bank);
+            //Allow default ddr_bank of -1; When CU is not connected to any ddr
+            // TODO: Removed this code and handle it using xrt::kernel
+            xma_xclbin_map2ddr(info.ip_ddr_mapping[d], &tmp1.default_ddr_bank, info.has_mem_groups);
+            //XMA now supports multiple DDR Banks per Kernel
+            tmp1.ip_ddr_mapping = info.ip_ddr_mapping[d]; 
+            for (int32_t c = 0; c < info.conn_axlf->m_count; c++)
+            {
+                if (info.conn_axlf->m_connection[c].m_ip_layout_index == (int32_t)d) {
+                    tmp1.CU_arg_to_mem_info.emplace(info.conn_axlf->m_connection[c].arg_index, info.conn_axlf->m_connection[c].mem_data_index);
                 }
-                /* Not to open context on all CUs
-                Will open during session_create
-                if (xclOpenContext(dev_tmp1.handle, info.uuid, d, true) != 0) {
-                    xma_logmsg(XMA_ERROR_LOG, XMAAPI_MOD, "Failed to open context to this CU\n");
-                    return false;
-                }
-                */
+            }
+            if (tmp1.default_ddr_bank < 0) {
+                xma_logmsg(XMA_WARNING_LOG, XMAAPI_MOD, "\tCU# %d - %s - default DDR bank: NONE", d, (char*)tmp1.name);
+            }
+            else {
+                xma_logmsg(XMA_DEBUG_LOG, XMAAPI_MOD, "\tCU# %d - %s - default DDR bank:%d", d, (char*)tmp1.name, tmp1.default_ddr_bank);
             }
         }
-
-        std::bitset<MAX_XILINX_KERNELS> cu_mask;
-        uint64_t base_addr1 = 0;
-        uint32_t cu_index_ert = 0;
-        for (uint32_t d1 = 0; d1 < info.number_of_hardware_kernels; d1++) {
-            base_addr1 = dev_tmp1.kernels[d1].base_address;
-            //uint64_t cu_mask = 1;
-            cu_mask.reset();
-            cu_mask.set(0);
-            cu_index_ert = 0;
-
-            for (uint32_t d2 = 0; d2 < info.number_of_hardware_kernels; d2++) {
-                if (base_addr1 == info.cu_addrs_sorted[d2]) {
-                    break;
-                }
-                cu_mask = cu_mask << 1;
-                cu_index_ert++;
-                /*
-                if (d1 != d2) {
-                    if (dev_tmp1.kernels[d2].base_address < base_addr1) {
-                        cu_mask = cu_mask << 1;
-                        cu_index_ert++;
-                    }
-                }
-                */
+        dev_tmp1.number_of_hardware_kernels = dev_tmp1.kernels.size();
+        uint32_t num_soft_kernels = 0;
+        //Handle soft kernels just like another hardware IP_Layout kernel
+        //soft kernels to follow hardware kernels. so soft kenrel index will start after hardware kernels
+        auto soft_kernels = xrt_core::xclbin::get_softkernels(xclbin_ax);
```

Comment:
Is this code before or after the xclbin has been loaded on the device?  If after, do we need to parse the xclbin, or will device_query do it?  E.g. check core/common/device.cpp:~179  where it gets the soft kernels.

Reply:
Noted down this comment, will check this logic in the next PR.